File: components/org.wso2.carbon.siddhi.editor.core/src/main/java/org/wso2/carbon/siddhi/editor/core/util/eventflow/EventFlow.java
/*
 * Copyright (c) 2018, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
 *
 * WSO2 Inc. licenses this file to you under the Apache License,
 * Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package org.wso2.carbon.siddhi.editor.core.util.eventflow;

import org.json.JSONArray;
import org.json.JSONObject;
import org.wso2.carbon.siddhi.editor.core.util.eventflow.constants.EdgeType;
import org.wso2.carbon.siddhi.editor.core.util.eventflow.constants.NodeType;
import org.wso2.carbon.siddhi.editor.core.util.eventflow.info.AggregationInfo;
import org.wso2.carbon.siddhi.editor.core.util.eventflow.info.FunctionInfo;
import org.wso2.carbon.siddhi.editor.core.util.eventflow.info.PartitionInfo;
import org.wso2.carbon.siddhi.editor.core.util.eventflow.info.PartitionTypeInfo;
import org.wso2.carbon.siddhi.editor.core.util.eventflow.info.QueryInfo;
import org.wso2.carbon.siddhi.editor.core.util.eventflow.info.SinkInfo;
import org.wso2.carbon.siddhi.editor.core.util.eventflow.info.SourceInfo;
import org.wso2.carbon.siddhi.editor.core.util.eventflow.info.StreamInfo;
import org.wso2.carbon.siddhi.editor.core.util.eventflow.info.TableInfo;
import org.wso2.carbon.siddhi.editor.core.util.eventflow.info.TriggerInfo;
import org.wso2.carbon.siddhi.editor.core.util.eventflow.info.WindowInfo;

import java.util.ArrayList;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * Obtains a SiddhiAppMap instance and generate's a JSON with a predefined format for the graph view in SP Editor UI.
 */
public class EventFlow {

    private static final String APP_NAME = "appName";
    private static final String APP_DESCRIPTION = "appDescription";
    private static final String TYPE = "type";
    private static final String NAME = "name";
    private static final String ID = "id";
    private static final String DESCRIPTION = "description";
    private static final String PARENT = "parent";
    private static final String CHILD = "child";
    private static final String NODE = "nodes";
    private static final String EDGE = "edges";
    private static final String GROUP = "groups";
    private static final String CHILDREN = "children";
    private static final Pattern PATTERN = Pattern.compile("\\b[i|I][n|N]\\b\\s+(\\w+)");

    private SiddhiAppMap siddhiAppMap;

    private JSONObject eventFlowJSON = new JSONObject();

    private JSONArray nodes = new JSONArray();
    private JSONArray edges = new JSONArray();
    private JSONArray groups = new JSONArray();

    public EventFlow(SiddhiAppMap siddhiAppMap) {
        this.siddhiAppMap = siddhiAppMap;
        setEventFlowJSON();
    }

    /**
     * Main method that is called in the constructor to generate a JSON
     * object from the provided SiddhiAppMap object.
     */
    private void setEventFlowJSON() {
        eventFlowJSON.put(APP_NAME, siddhiAppMap.getAppName());
        eventFlowJSON.put(APP_DESCRIPTION, siddhiAppMap.getAppDescription());
        setNodes();
        setEdges();
        setGroups();
    }

    /**
     * Creates all the node objects and adds it to the nodes JSONArray
     * which is then added to the returned JSONObject.
     */
    private void setNodes() {
        // Create Trigger Nodes
        for (TriggerInfo trigger : siddhiAppMap.getTriggers()) {
            createNode(NodeType.TRIGGER, trigger.getId(), trigger.getName(), trigger.getDefinition());
        }

        // Create Stream Nodes
        for (StreamInfo stream : siddhiAppMap.getStreams()) {
            createNode(NodeType.STREAM, stream.getId(), stream.getName(), stream.getDefinition());
        }

        // Create Source Nodes
        for (SourceInfo source : siddhiAppMap.getSources()) {
            createNode(NodeType.SOURCE, source.getId(), source.getName(), source.getDefinition());
        }

        // Create Sink Nodes
        for (SinkInfo sink : siddhiAppMap.getSinks()) {
            createNode(NodeType.SINK, sink.getId(), sink.getName(), sink.getDefinition());
        }

        // Create Table Nodes
        for (TableInfo table : siddhiAppMap.getTables()) {
            createNode(NodeType.TABLE, table.getId(), table.getName(), table.getDefinition());
        }

        // Create Window Nodes
        for (WindowInfo window : siddhiAppMap.getWindows()) {
            createNode(NodeType.WINDOW, window.getId(), window.getName(), window.getDefinition());
        }

        // Create Aggregation Nodes
        for (AggregationInfo aggregation : siddhiAppMap.getAggregations()) {
            createNode(NodeType.AGGREGATION, aggregation.getId(), aggregation.getName(), aggregation.getDefinition());
        }

        // Create Function Nodes
        for (FunctionInfo function : siddhiAppMap.getFunctions()) {
            createNode(NodeType.FUNCTION, function.getId(), function.getName(), function.getDefinition());
        }

        // Create Query Nodes
        for (QueryInfo query : siddhiAppMap.getQueries()) {
            createNode(NodeType.QUERY, query.getId(), query.getName(), query.getDefinition());
        }

        // Create Partition Nodes
        for (PartitionInfo partition : siddhiAppMap.getPartitions()) {
            createNode(NodeType.PARTITION, partition.getId(), partition.getName(), partition.getDefinition());

            // Create Nodes For The Queries Inside The Partition
            for (QueryInfo query : partition.getQueries()) {
                createNode(NodeType.QUERY, query.getId(), query.getName(), query.getDefinition());
            }

            // Create Nodes For The Range & Value Partition Types Inside The Partition
            for (PartitionTypeInfo partitionType : partition.getPartitionTypes()) {
                createNode(NodeType.PARTITION_TYPE, partitionType.getId(), partitionType.getName(),
                        partitionType.getDefinition());
            }
        }

        // Add All The Created Nodes To The Event Flow
        eventFlowJSON.put(NODE, nodes);
    }

    /**
     * Creates a JSONObject that defines a node and adds it to the nodes JSONArray.
     *
     * @param type        Defines the type of node (Ex: stream, table, trigger etc.)
     * @param id          A unique value given to identify a particular node
     * @param name        The name of the node (not always unique - this is the name that will be displayed in the UI)
     * @param description A piece of the Siddhi code where the node (stream, table etc.) is defined
     */
    private void createNode(NodeType type, String id, String name, String description) {
        JSONObject jsonObject = new JSONObject();
        jsonObject.put(TYPE, type.getTypeAsString());
        jsonObject.put(ID, id);
        jsonObject.put(NAME, name);
        jsonObject.put(DESCRIPTION, description);
        nodes.put(jsonObject);
    }

    /**
     * Calls all the relevant functions to create the edge JSONObjects
     * and then add these JSONObjects to the edges JSONArray.
     */
    private void setEdges() {
        setSourceEdges();
        setSinkEdges();
        setAggregationEdges();
        setQueryEdges();
        setPartitionEdges();
        eventFlowJSON.put(EDGE, edges);
    }

    /**
     * Creates all the edge JSONObjects for all the Sources in the SiddhhiAppMap object.
     */
    private void setSourceEdges() {
        for (SourceInfo source : siddhiAppMap.getSources()) {
            createEdge(EdgeType.DEFAULT, source.getId(), source.getStreamId());
        }
    }

    /**
     * Creates all the edge JSONObjects for all the Sinks in the SiddhiAppMap object.
     */
    private void setSinkEdges() {
        for (SinkInfo sink : siddhiAppMap.getSinks()) {
            createEdge(EdgeType.DEFAULT, sink.getStreamId(), sink.getId());
        }
    }

    /**
     * Creates all the edge JSONObjects for all the aggregations in the SiddhiAppMap object.
     */
    private void setAggregationEdges() {
        for (AggregationInfo aggregation : siddhiAppMap.getAggregations()) {
            createEdge(EdgeType.DEFAULT, aggregation.getInputStreamId(), aggregation.getId());
        }
    }

    /**
     * Creates all the edge JSONObjects for all the queries in the SiddhiAppMap object.
     */
    private void setQueryEdges() {
        // Set Edges With Queries
        for (QueryInfo query : siddhiAppMap.getQueries()) {
            // For Each Input Stream Id In The Query
            for (String inputStreamId : query.getInputStreamIds()) {
                createEdge(EdgeType.DEFAULT, inputStreamId, query.getId());
            }

            // Create An Edge Between The Query And It's OutputStream
            createEdge(EdgeType.DEFAULT, query.getId(), query.getOutputStreamId());

            // For Every User Defined Function Used In The Query
            for (String functionId : query.getFunctionIds()) {
                createEdge(EdgeType.DEFAULT, functionId, query.getId());
            }

            // Search For The 'in' Keyword Inside The Query To `Create Dotted Lined Edge`
            Matcher matcher = PATTERN.matcher(query.getDefinition());
            while (matcher.find()) {
                String tableId = matcher.group(1);
                createEdge(EdgeType.DOTTED_LINE, tableId, query.getId());
            }
        }
    }

    /**
     * Creates all the edge JSONObjects for all the partitions in the SiddhiAppMap object.
     */
    private void setPartitionEdges() {
        // Set Edges With Partition Queries
        for (PartitionInfo partition : siddhiAppMap.getPartitions()) {
            // For Each Query In This Partition
            for (QueryInfo query : partition.getQueries()) {
                createEdgesForQueryInPartition(query, partition);
            }
        }
    }

    /**
     * Creates all the edges for a particular query inside a partition.
     *
     * @param query     The query in which the edges should be made to and from
     * @param partition The partition in which the given query belongs to
     */
    private void createEdgesForQueryInPartition(QueryInfo query, PartitionInfo partition) {
        // For Each Input Stream In This Query
        for (String inputStreamId : query.getInputStreamIds()) {
            // Check Whether This Input Stream Is Partitioned Or Not
            String partitionedStreamId = null;
            for (PartitionTypeInfo partitionType : partition.getPartitionTypes()) {
                if (inputStreamId.equals(partitionType.getStreamId())) {
                    partitionedStreamId = partitionType.getId();
                    break;
                }
            }

            if (partitionedStreamId != null) {
                // If The Input Stream Is Partitioned
                createEdge(EdgeType.DEFAULT, inputStreamId, partitionedStreamId);
                createEdge(EdgeType.DEFAULT, partitionedStreamId, query.getId());
            } else {
                // If The Input Stream Is Not Partitioned
                createEdge(EdgeType.DEFAULT, inputStreamId, query.getId());
            }
        }

        // Connect The Query With Its Output Stream
        createEdge(EdgeType.DEFAULT, query.getId(), query.getOutputStreamId());

        // Connect All The Functions That Are Used By This Query
        for (String functionId : query.getFunctionIds()) {
            createEdge(EdgeType.DEFAULT, functionId, query.getId());
        }

        // Connects Any Tables That Are Used By The Queries Using The 'in' Keyword
        Matcher matcher = PATTERN.matcher(query.getDefinition());
        while (matcher.find()) {
            String tableId = matcher.group(1);
            createEdge(EdgeType.DOTTED_LINE, tableId, query.getId());
        }
    }

    /**
     * Creates a JSONObject that defines an edge and adds it to the edges JSONArray.
     *
     * @param edgeType The type of edge (can be either a 'arrow' or 'dotted-line')
     * @param parent   The Id of the parent node where the edge starts from
     * @param child    The Id of the child node where the edge should point towards
     */
    private void createEdge(EdgeType edgeType, String parent, String child) {
        JSONObject jsonObject = new JSONObject();
        jsonObject.put(TYPE, edgeType.getTypeAsString());
        jsonObject.put(PARENT, parent);
        jsonObject.put(CHILD, child);
        edges.put(jsonObject);
    }

    /**
     * Creates all the group objects and adds it to the groups JSONArray
     * which is then added to the returned JSONObject.
     */
    private void setGroups() {
        // Create A Group For Every Partition
        for (PartitionInfo partition : siddhiAppMap.getPartitions()) {
            createGroup(partition.getId(), partition.getName(), partition.getQueries(), partition.getPartitionTypes());
        }
        eventFlowJSON.put(GROUP, groups);
    }

    /**
     * Creates a JSONObject that defines a group and adds it to the groups JSONArray.
     * Groups are only meant for partitions as they define what streams/queries and partition types that
     * belong to a particular partition.
     *
     * @param id             The unique Id of the partition which the group belongs to
     * @param name           The name of the partition (i.e. the name to be displayed in the UI)
     * @param queries        The list of queries defined inside a particular partition
     * @param partitionTypes The list of nodes that define's how a particular stream is partitioned by
     */
    private void createGroup(String id, String name, List<QueryInfo> queries, List<PartitionTypeInfo> partitionTypes) {
        // Create A Group To Define A Partition
        JSONObject jsonObject = new JSONObject();
        jsonObject.put(ID, id);
        jsonObject.put(NAME, name);

        List<String> children = new ArrayList<>();
        // Add All The Queries To The Partition As Children
        for (QueryInfo query : queries) {
            children.add(query.getId());
            if (query.getOutputStreamId().substring(0, 1).equals("#")) {
                children.add(query.getOutputStreamId());
            }
        }
        // Add All The Partitioned Values As Children
        for (PartitionTypeInfo partitionType : partitionTypes) {
            children.add(partitionType.getId());
        }
        jsonObject.put(CHILDREN, children.toArray());

        // Add The Created JSONObject To The Groups JSONArray
        groups.put(jsonObject);
    }

    // Getters
    public JSONObject getEventFlowJSON() {
        return eventFlowJSON;
    }

}

File: components/org.wso2.carbon.siddhi.editor.core/src/main/resources/web/js/event-flow/event-flow.js
/**
 * Copyright (c) 2018, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
 *
 * WSO2 Inc. licenses this file to you under the Apache License,
 * Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

define(['require', 'log', 'lodash', 'jquery', 'alerts', 'd3', 'dagre_d3', 'overlay_scroller'],
    function (require, log, _, $, alerts, d3, dagreD3, overlayScroller) {

        /**
         * Create an instance of the EventFlow class in JavaScript.
         * This class is what is used to connect to the back-end and generate a graph in the front end design view.
         *
         * @param designView
         * @constructor Sets the default variables
         */
        var EventFlow = function (designView) {
            this.$designView = designView;
            this.$designViewData = this.$designView.find('.design-container-data');
            this.$siddhiAppName = this.$designView.find('.siddhi-app-name');
            this.$siddhiAppDescription = this.$designView.find('.siddhi-app-description');
            this.$graphView = this.$designView.find('.graph-container');
            this.$loadingScreen = this.$designView.find('.loading-screen');
            this.$siddhiGraph = this.$designView.find('.siddhi-graph');

            this.url = window.location.protocol + "//" + window.location.host + "/editor/event-flow";
            // The render options contains any settings/options for rendering the graph,
            // this has been added here separately for easier access if the style of the
            // graph generated needs to be changed.
            var defaultNodeStyle = {
                labelType: "html",
                rx: 0,
                ry: 0,
                padding: 0,
            };
            this.renderOptions = {
                nodeOptions: {
                    stream: {
                        name: "stream",
                        nodeStyle: defaultNodeStyle,
                        cssClass: "indicator stream-colour"
                    },
                    source: {
                      name: "source",
                      nodeStyle: defaultNodeStyle,
                      cssClass: "indicator source-colour"
                    },
                    sink: {
                        name: "sink",
                        nodeStyle: defaultNodeStyle,
                        cssClass: "indicator sink-colour"
                    },
                    table: {
                        name: "table",
                        nodeStyle: defaultNodeStyle,
                        cssClass: "indicator table-colour"
                    },
                    window: {
                        name: "window",
                        nodeStyle: defaultNodeStyle,
                        cssClass: "indicator window-colour"
                    },
                    trigger: {
                        name: "trigger",
                        nodeStyle: defaultNodeStyle,
                        cssClass: "indicator trigger-colour"
                    },
                    aggregation: {
                        name: "aggregation",
                        nodeStyle: defaultNodeStyle,
                        cssClass: "indicator aggregation-colour"
                    },
                    function: {
                        name: "function",
                        nodeStyle: defaultNodeStyle,
                        cssClass: "indicator function-colour"
                    },
                    query: {
                        name: "query",
                        nodeStyle: defaultNodeStyle,
                        cssClass: "indicator query-colour"
                    },
                    partition: {
                        name: "partition",
                        nodeStyle: {
                            labelType: "html",
                            clusterLabelPos: "top",
                            style: "fill: #434343"
                        },
                        cssClass: "partition"
                    },
                    partition_type: {
                        name: "partition-type",
                        nodeStyle: defaultNodeStyle,
                        cssClass: "indicator partition-type-colour"
                    }
                },
                edgeOptions: {
                    default: {
                        name: "default",
                        edgeStyle: {
                            arrowheadStyle: "fill: #bbb",
                            lineInterpolate: "basis"
                        }
                    },
                    dotted_line: {
                        name: "dotted-line",
                        edgeStyle: {
                            arrowheadStyle: "fill: #bbb",
                            style: "stroke-dasharray: 5, 5;"
                        }
                    }
                }
            };

            var self = this;
            $(window).on('resize', function () {
                if (self.isVisible()) {
                    self.graphResize();
                }
            });

            this.$graphView.overlayScrollbars({className: "os-theme-light"});
        };

        /**
         * Does an AJAX call to the backend MSF4J service that converts the 'siddhiCode'
         * to a specific JSON format to be displayed in the UI.
         *
         * @param siddhiCode The Siddhi App as a string
         * @returns A JSON with a response status, and the JSON value returned by the back-end
         */
        EventFlow.prototype.fetchJSON = function (siddhiCode) {
            var self = this;
            var result = {};

            if (siddhiCode === null || siddhiCode === undefined || siddhiCode === "") {
                result = {status: "fail", errorMessage: "The Siddhi App Cannot Be Empty"};
            } else {
                // Remove Single Line Comments
                var regexStr = siddhiCode.replace(/--.*/g, '');
                // Remove Multi-line Comments
                var regexStr = regexStr.replace(/\/\*(.|\s)*?\*\//g, '');
                var regex = /^\s*@\s*app\s*:\s*name\s*\(\s*["|'](.*)["|']\s*\)\s*@\s*app\s*:\s*description\s*\(\s*["|'](.*)["|']\s*\)\s*$/gi;
                var match = regex.exec(regexStr);

                if (match !== null) {
                    result = {
                        status: "success",
                        responseJSON: {
                            appName: match[1],
                            appDescription: match[2],
                            nodes: [],
                            edges: [],
                            groups: []
                        }
                    };
                } else {
                    fetch(siddhiCode);
                }

            }

            function fetch(code) {
                $.ajax({
                    type: "POST",
                    url: self.url,
                    data: window.btoa(code),
                    async: false,
                    success: function (response) {
                        result = {status: "success", responseJSON: response};
                    },
                    error: function (error) {
                        if (error.status === 400) {
                            result = {status: "fail", errorMessage: "Siddhi App Contains Errors"};
                        } else {
                            result = {status: "fail", errorMessage: "Internal Server Error Occurred"};
                        }
                    }
                });
            }

            return result;
        };

        /**
         * Renders a graph in the design view based on the information of the structure
         * of the graph in the 'data' argument.
         *
         * @param data A JSON object that defines the graph that is to be rendered in the UI
         */
        EventFlow.prototype.render = function (data) {
            var self = this;

            if (data === null || data === undefined || data === {}) {
                log.error("The argument sent to the render method is null/undefined");
            } else {
                createGraph();
                self.$loadingScreen.hide();
            }

            function createGraph() {
                if (data.appName === null || data.appName === undefined || data.appName === "") {
                    self.$siddhiAppName.html("SiddhiApp");
                } else {
                    self.$siddhiAppName.html(data.appName);
                }

                if (data.appDescription === null || data.appDescription === undefined || data.appDescription === "") {
                    self.$siddhiAppDescription.html("Description of the plan");
                } else {
                    self.$siddhiAppDescription.html(data.appDescription);
                }

                var nodeOptions = self.renderOptions.nodeOptions;
                var edgeOptions = self.renderOptions.edgeOptions;

                // Create an instance of the dagreD3 graph.
                var graph = new dagreD3.graphlib.Graph({compound: true}).setGraph({});
                // This makes sure the graph grows from left-to-right.
                graph.graph().rankDir = "LR";

                // Set the nodes of the graph
                data.nodes.forEach(function (node) {
                    var html;
                    var nodeStyle;

                    switch (node.type) {
                        case nodeOptions.stream.name:
                            html = "<div class='node-content' title='" + node.description + "'>"
                                + "<span class='" + nodeOptions.stream.cssClass + "'></span>"
                                + "<span class='nodeLabel'>" + node.name + "</span>"
                                + "</div>";
                            // The _.clone() method must be used to avoid passing by reference.
                            nodeStyle = _.clone(nodeOptions.stream.nodeStyle);
                            nodeStyle.label = html;
                            break;
                        case nodeOptions.source.name:
                            html = "<div class='node-content' title='" + node.description + "'>"
                                + "<span class='" + nodeOptions.source.cssClass + "'></span>"
                                + "<span class='nodeLabel'>" + node.name + "</span>"
                                + "</div>";
                            nodeStyle = _.clone(nodeOptions.source.nodeStyle);
                            nodeStyle.label = html;
                            break;
                        case nodeOptions.sink.name:
                            html = "<div class='node-content' title='" + node.description + "'>"
                                + "<span class='" + nodeOptions.sink.cssClass + "'></span>"
                                + "<span class='nodeLabel'>" + node.name + "</span>"
                                + "</div>";
                            nodeStyle = _.clone(nodeOptions.sink.nodeStyle);
                            nodeStyle.label = html;
                            break;
                        case nodeOptions.table.name:
                            html = "<div class='node-content' title='" + node.description + "'>"
                                + "<span class='" + nodeOptions.table.cssClass + "'></span>"
                                + "<span class='nodeLabel'>" + node.name + "</span>"
                                + "</div>";
                            nodeStyle = _.clone(nodeOptions.table.nodeStyle);
                            nodeStyle.label = html;
                            break;
                        case nodeOptions.window.name:
                            html = "<div class='node-content' title='" + node.description + "'>"
                                + "<span class='" + nodeOptions.window.cssClass + "'></span>"
                                + "<span class='nodeLabel'>" + node.name + "</span>"
                                + "</div>";
                            nodeStyle = _.clone(nodeOptions.window.nodeStyle);
                            nodeStyle.label = html;
                            break;
                        case nodeOptions.trigger.name:
                            html = "<div class='node-content' title='" + node.description + "'>"
                                + "<span class='" + nodeOptions.trigger.cssClass + "'></span>"
                                + "<span class='nodeLabel'>" + node.name + "</span>"
                                + "</div>";
                            nodeStyle = _.clone(nodeOptions.trigger.nodeStyle);
                            nodeStyle.label = html;
                            break;
                        case nodeOptions.aggregation.name:
                            html = "<div class='node-content' title='" + node.description + "'>"
                                + "<span class='" + nodeOptions.aggregation.cssClass + "'></span>"
                                + "<span class='nodeLabel'>" + node.name + "</span>"
                                + "</div>";
                            nodeStyle = _.clone(nodeOptions.aggregation.nodeStyle);
                            nodeStyle.label = html;
                            break;
                        case nodeOptions.function.name:
                            html = "<div class='node-content' title='" + node.description + "'>"
                                + "<span class='" + nodeOptions.function.cssClass + "'></span>"
                                + "<span class='nodeLabel'>" + node.name + "</span>"
                                + "</div>";
                            nodeStyle = _.clone(nodeOptions.function.nodeStyle);
                            nodeStyle.label = html;
                            break;
                        case nodeOptions.query.name:
                            html = "<div class='node-content' title='" + node.description + "'>"
                                + "<span class='" + nodeOptions.query.cssClass + "'></span>"
                                + "<span class='nodeLabel'>" + node.name + "</span>"
                                + "</div>";
                            nodeStyle = _.clone(nodeOptions.query.nodeStyle);
                            nodeStyle.label = html;
                            break;
                        case nodeOptions.partition.name:
                            html = "<div class='" + nodeOptions.partition.cssClass
                                + "' title='" + node.description + "'>"
                                + node.name
                                + "</div>";
                            nodeStyle = _.clone(nodeOptions.partition.nodeStyle);
                            nodeStyle.label = html;
                            break;
                        case nodeOptions.partition_type.name:
                            html = "<div class='node-content' title='" + node.description + "'>"
                                + "<span class='" + nodeOptions.partition_type.cssClass + "'></span>"
                                + "<span class='nodeLabel'>" + node.name + "</span>" + "</div>";
                            nodeStyle = _.clone(nodeOptions.partition_type.nodeStyle);
                            nodeStyle.label = html;
                            break;
                    }

                    // Set the node
                    graph.setNode(node.id, nodeStyle);

                });

                // Set the edges of the graph
                data.edges.forEach(function (edge) {
                    var edgeStyle;
                    if (edge.type === "dotted-line") {
                        // This makes the lines dotted
                        edgeStyle = _.clone(edgeOptions.dotted_line.edgeStyle);
                    } else {
                        edgeStyle = _.clone(edgeOptions.default.edgeStyle);
                    }

                    // Set the edge
                    graph.setEdge(edge.parent, edge.child, edgeStyle);
                });

                // Set the groups of the graph
                // NOTE - The nodes and edges should be created first before the groups can be created
                data.groups.forEach(function (group) {
                    group.children.forEach(function (child) {
                        graph.setParent(child, group.id);
                    });
                });

                var render = new dagreD3.render();
                var graphId = "#" + self.$siddhiGraph.attr("id");
                render(d3.select(graphId + " g"), graph);

                resizeGraph(self.$siddhiGraph);

                function resizeGraph(svg) {
                    var inner = svg.find('g');
                    svg.attr("width", inner.get(0).getBoundingClientRect().width + 60);
                    svg.attr("height", inner.get(0).getBoundingClientRect().height + 60);
                }
            }
        };

        EventFlow.prototype.graphResize = function () {
            var self = this;

            resizeGraph();
            centerSVG();

            function resizeGraph() {
                var height = self.$designView.height() - self.$designViewData.height() - 20;
                self.$graphView.height(height);
            }

            function centerSVG() {
                var graphWidth = parseInt(self.$siddhiGraph.attr("width"));
                var graphHeight = parseInt(self.$siddhiGraph.attr("height"));
                var width = self.$graphView.width();
                var height = self.$graphView.height();

                var left = diff(width, graphWidth) / 2;
                var top = diff(height, graphHeight) / 2;
                self.$siddhiGraph.attr("transform", "translate(" + left + "," + top + ")");

                function diff(divValue, graphValue) {
                    if (divValue > graphValue) {
                        return (divValue - graphValue);
                    } else {
                        return 0;
                    }
                }
            }
        };

        EventFlow.prototype.isVisible = function () {
            return this.$designView.is(':visible');
        };

        /**
         * Clears the design view of all it's contents when called.
         */
        EventFlow.prototype.clearContent = function () {
            this.$siddhiAppName.empty();
            this.$siddhiAppDescription.empty();
            this.$loadingScreen.show();
            this.$siddhiGraph.empty();
            this.$siddhiGraph.html('<g></g>');
        };

        /**
         * Display's a warning using the AlertsManager.
         *
         * @param message The content to be displayed in the alert
         */
        EventFlow.prototype.alert = function (message) {
            alerts.warn(message);
        };

        return EventFlow;

    });
